### DISCUSSION TITLE
OTA Update with local files for arbitrary devices? (yet another question)

### DISCUSSION THREAD

**Discussion by grafalex82** [Koenkk/zigbee2mqtt#10032] [Devices]
I am developing my own device, and going to add OTA update feature. Basically I have everything working already, and currently testing by setting up a local web server, according to instructions [here](https://github.com/Koenkk/zigbee2mqtt/issues/3118#issuecomment-607077467).

In my opinion this is very inconvenient:
- This is forcing users of my DIY device to have a debug build of z2m
- This is forcing users to set up a web server (even a simple one, such as python http server)
- requires changes in z2m/herdsman code
- once you are pointing z2m/herdsman to the local web server, you no longer have updates from 'official' [Zigbee-OTA repo](https://github.com/Koenkk/zigbee-OTA), and therefore have no updates.

It is possible to fork the repo, and upload firmwares to the github, but the concern about having to patch z2m/herdsman, as well as other extra steps needed still valid.

The aim of this message is to start a public discussion on the best way to implement local OTA updates. I think since Z2M allows custom converters placed locally, why can't it allow local OTA firmware files?

Here is how I see it. I think it would be convenient for users to put firmwares in the data directory, next to configuration.yml and custom converters. I think having a database (like [index.json](https://github.com/Koenkk/zigbee-OTA/blob/master/index.json)) get things overcomplicated - all the information about OTA firmware could be easily extracted from the firmware file.

Here I am speculating on how users could define information for local firmware files.

Option 1:
The file name is defined in converter file. For example

```
const device = {
    zigbeeModel: ['Hello Zigbee Switch'],
    model: 'Hello Zigbee Switch',
...
    ota: ota.localOTA,
    firmware: 'HelloZigbee.ota',
}
```

HelloZigbee.ota is searched in the same directory as converter js file (which in turn is searched next to configuration.yml)

Option 2:
The file name is defined in configuration.yml, such as

```
external_converters:
  - myswitch.js
devices:
  '0x00158d0002b501a7':
    friendly_name: TestSwitch
    ota: 'HelloZigbee.ota'
```

**What do you think about one of these approaches?**

In the meantime I managed to pass 'Check Update Available' stage just by adding a code in my external converter file
```
async function getImageMeta(current, logger, device) {
    logger.debug(`My getImageMeta()`);
    logger.debug(`device='${JSON.stringify(device)}'`);
    logger.debug(`current='${JSON.stringify(current)}'`);

    const modelId = device.modelID;

    const fileName = '/home/grafalex/1/zigbee2mqtt/data/HelloZigbee.ota';
    const buffer = fs.readFileSync(fileName);
    const parsed = otacommon.parseImage(buffer);

    const hash = crypto.createHash('sha512');
    hash.update(buffer);

    const ret = {
        fileVersion: parsed.header.fileVersion,
        fileSize: parsed.header.totalImageSize,
        url: fileName,
        sha512: hash.digest('hex'),
    };

    logger.debug(`My getImageMeta(): ret='${JSON.stringify(ret)}'`);
    return ret;
}

const MyOta = {

    isUpdateAvailable: async (device, logger, requestPayload=null) => {
        logger.debug(`My isUpdateAvailable()`);
        return otacommon.isUpdateAvailable(device, logger, otacommon.isNewImageAvailable, requestPayload, getImageMeta);
    },

    updateToLatest: async (device, logger, onProgress) => {
        logger.debug(`My updateToLatest()`);
        return otacommon.updateToLatest(device, logger, onProgress, otacommon.getNewImage, getImageMeta);
    }
}
```

There are 3 problems with this code:
1) External converter is executed in a vm.runInNewContext() session, and sub-interpreter does not actually know the path to the converter, and therefore can't calculate the path to OTA firmware next to it
2) herdsman and OTA code does not have access to Z2M configuration, so I am not sure how to pass any information from configuration.yml to herdsman's OTA code (please advice)
3) OTA code does not work with local files, and expects only external URLs. Currently it looks like a technical impediment, but perhaps there is an important reason behind this.

Thoughts?

**Comment by kirovilya** [comment]
Agree. The topic of updating devices with their own firmware has already been raised in the discussions of the Russian community, but it remained unresolved.

I propose a slightly different way to expand the OTA functionality: add the ability to specify one or more OTA-index files and OTA-servers in the settings. It will be similar to specifying external device converters.
If a file is specified, then it must have the same structure as index.json so that its processing coincides with the work of the general OTA-index server.
In such an external file, you can specify the local firmware files, their versions and device models for which they are suitable.
As with converters, local files should take precedence.

Another feature I see is the ability to update firmware with a lower or the same version. This is sometimes needed to return to a previous state.
You can add some "force" attribute for such cases.

**Reply by grafalex82** [reply]
Well, this is looks doable, and solves necessity of patching z2m and setting up debug z2m build. Perhaps this is a good starting point.

However, this still requires user to set up own http server and maintain index.json.

**Reply by kirovilya** [reply]
no, you don't need a server, you only need an index-file (with an arbitrary name), the local path to which is specified in the settings

**Reply by kirovilya** [reply]
and in this file, links to firmware can also be local - indicate the path to the file

**Reply by grafalex82** [reply]
Why do we need an index file, if all the information can be obtained from the firmware binary? Moreover this is an additional step for the user to update the index file when a new firmware is available.

**Reply by kirovilya** [reply]
not all information can be obtained from the firmware file. more precisely, not always. for example, in some firmwares for xiaomi / gledopto it is impossible to determine which device the firmware is intended for. so the model is manually specified in the file.
I suggested leaving the local index file in order not to fundamentally change the approach to working with firmware.

**Reply by grafalex82** [reply]
Ok, makes sense. But this means that z2m must provide a tool for adding firmwares to the index (which may be complicated if z2m is running in a docker container).

Alternatively, this index could be a yaml or json file with a list of firmwares, and set of optional fields to clarify the match. For example (syntax mistakes possible):

```
firmwares:
- file: Firmware1.ota    # All the information is parsed from the file
- file: Firmware2.ota    # This firmware could be applied only for devices produced by manufacturer 0x1234
  manufacturerCode: 0x1234
- file: Firmware4.ota    # This firmware could be applied only to a device with all the fields matched 
  deviceName: "My test device"
  manufacturerCode: 0x1234
  imageType: 2
  modelId: "abcdefgh"
```

This file could be easily prepared by human, and firmware producer could just provide the firmware with a yaml snippet, that user can just apply to their configuration file.

**Reply by grafalex82** [reply]
After having an offline conversation with @kirovilya I propose the following plan:
1. Add a possibility for a local OTA index file. This step has the following benefits:
- no need to patch z2m/herdsman code to point it to an alternate OTA index file
- Add new firmwares without uploading them to a global zigbee-OTA index
- At the same time changes in global zigbee-OTA repo will be still available
- Override firmwares for a specific device model/device type

This part is done in https://github.com/Koenkk/zigbee-herdsman-converters/pull/3495 and https://github.com/Koenkk/zigbee2mqtt/pull/10141

This step does not assumes local firmwares - firmware files still have to be downloaded from an http server

2. Support local firmware files. I think 'url' field in the index file could also have path to the local file. Or it could be a different field. Path to the firmware file can be either absolute, or relative to configuration.yaml location.

3. At the same time, this approach will still require index file concept, as well running add.js (which also needs to be patched). 

During this stage we can think about further simplification of the approach, such as I described in the previous post.
